### PR TITLE
Profile Update Component

### LINKED ISSUE #123: Profile updation component
## Description
As a User,
I need Profile Updation page,
so that I can upgrade my profile information such as profile pic, email address, phone number.

### Update [Required]
- [ ] Profile Updation UI
- [ ] Profile Updation Backend Changes
- [ ] Backend UI Integration
- [ ] Unit test coverage (both front end and backend) > 80%  
- [ ] Passing Travis.cli

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentor.

## Estimation
9 -11 hours

### PR DESCRIPTION
**Pull request by lunayach**
### Description
Profile Update Component

Fixes #123

### Type of Change:

- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Describe the tests you ran to verify your changes. Provide instructions or GIFs so we can reproduce. List any relevant details for your test.
Run `ng test` and `npm test`

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: src/app/app.routing.ts [hunk 1]
```diff
@@ -27,132 +27,139 @@ import { StageCompleteComponent } from './stage-complete/stage-complete.componen
...
+  },
+  {
+    path: 'register',
+    component: RegisterComponent,
+    canActivate: [UnauthenticatedGuard]
+  },
+  {
+    path: 'forgot',
+    component: ForgotPasswordComponent,
+    // canActivate: [UnauthenticatedGuard]
```
Comment (Raxerz) [inline src/app/app.routing.ts]: Why is this commented out?

File: src/app/navbar/navbar.component.html [hunk 2]
```diff
@@ -8,6 +8,7 @@
 
     <ul class="nav text-center">
       <div class="profile">
+        <div style="text-align: right"><button mat-button (click)="editUser()"> <mat-icon>edit</mat-icon></button></div>
```
Comment (Raxerz) [inline src/app/navbar/navbar.component.html]: please move the inline style to the scss file

File: src/tests/Services/update-user.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { TestBed, inject } from '@angular/core/testing';
```
Comment (Raxerz) [inline src/tests/Services/update-user.service.spec.ts]: Can it have a test case to check if the UI is updated?

File: src/app/user-update/user-update.component.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import { SharedDataService } from '../services/shared.data.service';
+
+@Component({
+  selector: 'app-user-update',
+  templateUrl: './user-update.component.html',
+  styleUrls: ['./user-update.component.scss']
+})
+export class UserUpdateComponent implements OnInit {
+
+  username: string;
```
Comment (shreyans29) [inline src/app/user-update/user-update.component.ts]: Make a model of it and then create an object.

File: src/tests/Services/update-user.service.spec.ts (hunk 3 above)
Comment (lunayach) [reply src/tests/Services/update-user.service.spec.ts]: Done.

**Comment by lunayach** [comment]
demo

File: data/codes.json [hunk 5]
```diff
@@ -175,7 +175,11 @@
         "PCS005": {
             "code": "PCS005",
             "message": "PCS005 : Verification mail sent, please check your mail"
-        }        
+        }      ,
```
Comment (Raxerz) [inline data/codes.json]: spacing, please keep it uniform with the rest of the contents in this file

File: src/app/navbar/navbar.component.ts [hunk 6]
```diff
@@ -67,5 +68,16 @@ export class NavbarComponent implements OnInit {
...
         this.infoPop.emit();
     }
+
+    editUser(): void {
+      const dialogConfig = new MatDialogConfig();
+      dialogConfig.data = {
+        profilePicture: this.proPic,
+        username: this.username
+      };
+      dialogConfig.width = '600px';
```
Comment (Raxerz) [inline src/app/navbar/navbar.component.ts]: is this responsive?

File: src/app/services/update-user.service.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+
+  constructor(private _apiService: APIService) {
+  }
+
+  public updateUserDetails(body: Object): Observable<any> {
+    return this._apiService.patch(this._updateAPI, body);
+  }
+
+}
```
Comment (Raxerz) [inline src/app/services/update-user.service.ts]: new line

File: routes/update.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    }
+});
+
+
+
+
+
+
+module.exports = router;
```
Comment (Raxerz) [inline routes/update.js]: new line

File: src/app/user-update/user-update.component.ts [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    });
+    this._canvas = document.getElementById('canvas');
+    this._stage = this._canvas.getContext('2d');
+    this._img = new Image();
+    this._img.src = this._source;
+    this._stage.drawImage(this._img, 0, 0);
+  }
+
+
+}
```
Comment (Raxerz) [inline src/app/user-update/user-update.component.ts]: new line

File: src/app/user-update/user-update.component.ts (hunk 4 above)
Comment (lunayach) [reply src/app/user-update/user-update.component.ts]: All of these 3 keys in the same object may not be useful given the first name and the last name key are dependent on the former.

File: src/app/navbar/navbar.component.ts (hunk 6 above)
Comment (lunayach) [reply src/app/navbar/navbar.component.ts]: Surprisingly, it turned out to be responsive!! I think this width property is = max-width here.

File: data/codes.json (hunk 5 above)
Comment (lunayach) [reply data/codes.json]: Done.

File: src/app/services/update-user.service.ts (hunk 7 above)
Comment (lunayach) [reply src/app/services/update-user.service.ts]: Done.

File: routes/update.js (hunk 8 above)
Comment (lunayach) [reply routes/update.js]: Done.

**Comment by shreyans29** [comment]
The new profile update component looks neat.. I was wondering where is the activity indicator ? 
It was supposed to be like a footer below. Also Click to upload has to be in center.. It is little drifted towards right.

File: src/app/navbar/navbar.component.ts (hunk 6 above)
Comment (shreyans29) [inline src/app/navbar/navbar.component.ts]: Is there no other way to mention this.. you should be using ts files for css properties as last resort..

File: src/app/user-update/user-update.component.ts (hunk 4 above)
Comment (shreyans29) [reply src/app/user-update/user-update.component.ts]: okay makes sense, but try doing that where ever you can like in the strides against malaria game.

File: src/app/navbar/navbar.component.ts (hunk 6 above)
Comment (lunayach) [reply src/app/navbar/navbar.component.ts]: @shreyans29 last resort would be to make a config file. But with having only this entry in it.

**Comment by Raxerz** [comment]
@lunayach Add a default image if there is no profile picture set previously. 
pro-pic

**Comment by lunayach** [comment]
@Raxerz added the default message and first name validation.

**Comment by Raxerz** [comment]
@lunayach The alignment is off now, the default profile pic also seems to be cutoff.
profileupd

File: routes/update.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+        }).catch(function() {
+            return res.status(500).json({error: errorCode.PCE028.message, code: errorCode.PCE028.code});
+        });
+    } else {
+        return res.status(400).json({error: errorCode.PCE029.message, code: errorCode.PCE029.code});
+
+    }
+});
+module.exports = router;
```
Comment (Raxerz) [inline routes/update.js]: new line

**Comment by paavininanda** [comment]
Also the edit button was supposed to be re positioned. @Raxerz @shreyans29 @Buddhiprabha We didn't decide anything for that? I think this button can be placed exactly to the right of the name. What do you guys think ?

**Comment by lunayach** [comment]
Have done the required changes!

File: routes/update.js (hunk 10 above)
Comment (shreyans29) [reply routes/update.js]: I suggest you configure your editor to do this. Atom has an option to add the extra new line if its not there when you save.

File: src/assets/languages/english.json [hunk 11]
```diff
@@ -67,6 +67,29 @@
         "registrationFail": "Registration unsuccesful"
       }
     },
+    "userUpdate": {
+      "messages": {
+        "heading": "Update your profile!",
+        "buttonMessage": "UPDATE",
+        "hoverProfile": "Click to upload",
+        "firstUpload": "You don't seem to have. Upload one :)",
```
Comment (Raxerz) [inline src/assets/languages/english.json]: I don't think this line is required.

File: src/assets/languages/english.json [hunk 12]
```diff
@@ -67,6 +67,29 @@
         "registrationFail": "Registration unsuccesful"
       }
     },
+    "userUpdate": {
+      "messages": {
+        "heading": "Update your profile!",
+        "buttonMessage": "UPDATE",
+        "hoverProfile": "Click to upload",
```
Comment (Raxerz) [inline src/assets/languages/english.json]: Change 'Click to upload' to "Change profile picture", and on clicking a dialog appears asking the user if they want to upload or remove the picture.

Comment (lunayach) [reply src/assets/languages/english.json]: @Raxerz "first upload" message would serve as a default message when he is not having the profile picture already. For the add/delete functionality, another dialog would need to be implemented. Is it a priority?

Comment (Raxerz) [reply src/assets/languages/english.json]: @shreyans29 @paavininanda @Buddhiprabha Should we take it in this PR or should we create a separate issue for it?

File: src/assets/languages/english.json (hunk 11 above)
Comment (Raxerz) [reply src/assets/languages/english.json]: @shreyans29 @paavininanda Is this fine? I feel it sounds very informal. Please add your comments.

File: src/assets/languages/english.json (hunk 12 above)
Comment (Buddhiprabha) [reply src/assets/languages/english.json]: Yes I too think it is required. If this was not there in the original suggestions, shall we add it as another issue and assign to @lunayach? @Raxerz

Comment (Raxerz) [reply src/assets/languages/english.json]: @Buddhiprabha Yes lets add it as another issue.

Comment (shreyans29) [reply src/assets/languages/english.json]: Oops ! late to the discussion, but I would also suggest another Issue.

File: src/assets/languages/english.json (hunk 11 above)
Comment (paavininanda) [reply src/assets/languages/english.json]: Yea.. this can be removed.

**Comment by lunayach** [comment]
@Raxerz Removed that "first Upload" message thing. Also as expected, editProfile() is working from the new position of the edit button in "Brand Update" PR after rebasing.

File: test/test.js [hunk 13]
```diff
@@ -144,6 +145,21 @@ describe('Testing APIs', function() {
                     });
             });
 
+            it('Updates User Profile', function(done) {
```
Comment (paavininanda) [inline test/test.js]: @Raxerz Any idea why this might be failing ?

**Pull request merged**